packages/bridge-ui-v2/src/components/Bridge/Actions.svelte

```diff
@@ -0,0 +1,96 @@
+<script lang="ts">
+  import { t } from 'svelte-i18n';
+
+  import { Button } from '$components/Button';
+  import { Icon } from '$components/Icon';
+  import { isERC20 } from '$libs/token';
+  import { account, network } from '$stores';
+
+  import {
+    computingBalance,
+    destNetwork,
+    enteredAmount,
+    errorComputingBalance,
+    insufficientAllowance,
+    insufficientBalance,
+    recipientAddress,
+    selectedToken,
+    tokenBalance,
+  } from './state';
+
+  export let approve: () => Promise<void>;
+  export let bridge: () => Promise<void>;
+
+  let approving = false;
+  let bridging = false;
+
+  function onApproveClick() {
+    approving = true;
+    approve().finally(() => {
+      approving = false;
+    });
+  }
+
+  function onBridgeClick() {
+    bridging = true;
+    bridge().finally(() => {
+      bridging = false;
+    });
+  }
+
+  // TODO: feels like we need a state machine here
+
+  // Basic conditions so we can even start the bridging process
+  $: hasAddress = $recipientAddress || $account?.address;
+  $: hasNetworks = $network?.id && $destNetwork?.id;
+  $: hasBalance =
+    !$computingBalance && !$errorComputingBalance && $tokenBalance?.value && $tokenBalance?.value > BigInt(0);
+  $: canDoNothing = !hasAddress || !hasNetworks || !hasBalance || !$selectedToken || !$enteredAmount;
+
+  // Conditions for approve/bridge steps
+  $: isSelectedERC20 = $selectedToken && isERC20($selectedToken);
+  $: isTokenApproved = isSelectedERC20 && $enteredAmount && !$insufficientAllowance;
+
+  // Conditions to disable/enable buttons
+  $: disableApprove = canDoNothing || !$insufficientAllowance || approving;
+  $: disableBridge = canDoNothing || $insufficientAllowance || $insufficientBalance || bridging;
+
+  // General loading state
+  $: loading = approving || bridging;
+</script>
+
+<div class="f-between-center w-full gap-4">
+  {#if isSelectedERC20}
+    <Button
+      type="primary"
+      class="px-[28px] py-[14px] rounded-full flex-1"
+      disabled={disableApprove}
+      loading={approving}
+      on:click={onApproveClick}>
+      {#if approving}
+        <span class="body-bold">{$t('bridge.button.approving')}</span>
+      {:else if isTokenApproved}
+        <div class="f-items-center">
+          <Icon type="check" />
+          <span class="body-bold">{$t('bridge.button.approved')}</span>
+        </div>
+      {:else}
+        <span class="body-bold">{$t('bridge.button.approve')}</span>
+      {/if}
+    </Button>
+    <Icon type="arrow-right" />
+  {/if}
+
+  <Button
+    type="primary"
+    class="px-[28px] py-[14px] rounded-full flex-1"
+    disabled={disableBridge}
+    loading={bridging}
+    on:click={onBridgeClick}>
+    {#if bridging}
+      <span class="body-bold">{$t('bridge.button.bridging')}</span>
```

Comment:
Nice: Would be cool to see why the button is disabled, e.g. "insufficient funds". We have all the states for it anyway.

Reply:
It does, it says that under the entered amount, which is responsible for it. There are a few reasons why the button can be disabled:
1. Token requires allowance (only for ERC20), in which case this is informed to the user close to what's responsible, in this case the entered amount.
2. Entered amount is too high. The user has not enough balance to cover the cost of bridging. User is informed about this, same as previous one.
3. Base conditions are not  fulfilled: no token or chains has been selected, etc...

Reply:
What I meant display the reason inside the button text, I saw it was disabled when I switched to an account with 0 eth connected but it didn't specify why. I guess this is an edge case though as it did not re-trigger the input validation for amount that would display the correct message. Button was disabled, that's good enough.

Reply:
Yeah, we can ask Jane about this to include it in the design